### DISCUSSION TITLE
Invoke Machine vs Promise

### DISCUSSION THREAD

**Discussion by ayashjorden** [statelyai/xstate#1717] [Q&A]
Hey,
Wanted to ask about the Invoke block.
On the difference between using a `Machine` or an `async` function as the `src` value.

* Using a async function (which returns a Promise), the state machine forwards the last event to the function
* Using a Machine, needs to either forward or send an event to the invoke block.

See example: https://xstate.js.org/viz/?gist=cb2776b231c2bacac7fe49f40f74ec76
Chaining two states with an invoke block in each, results in the second invoke block getting the first invoke’s done event (done.invoke.faas1 in the above example).
What am I doing / understanding wrong?

Apparently there’s a difference between Invoke-ing a Machine and a function (async function)
* When using a function of type (c,e) => SOME_FUNCTION_BODY, XState forwards the previous event to the function.
* When using a Machine as the src property, the last event is not sent to the Invoke block, but there's a need for  xstate.send/xstate.forwardTo actions to send/forward someEvent to the Invoke block.

Appreciate any insights,
Yarden

**Comment by davidkpiano** [comment]
You might be thinking about this wrong, which is causing confusion.

With a function `(ctx, e) => ...`, you're not "forwarding the previous event" to an entity, you are creating an entity _from_ the previous `context` and `event` (whether you use them or not).

In other words, if I invoke a promise like:

```js
src: (context, event) => fetch(`/items/${event.id}`).then(...)
```

Then I'm not "forwarding" the event to the promise actor; I am _creating an actor using the event_.

The same is true for machines:

```js
src: (context, event) => createMachine(/* ... */)
```

So in short, you can also return machines from functions in `src`, which is provided the most recent event. But the most recent event is _not_ forwarded (sent) to the invoked actor; that's a misconception.

**Reply by ayashjorden** [reply]
Indeed confusing....

Then what is the difference between
```typescript
src: (c,e) => fetch(`/items/${event.id}`).then(...)
```

and `src: "fetch"` (where `fetch` is declared somewhere as `async function fetch(c,e) {function-body-here}`.


Re `Machines`, in my project, I don't use `src: functionThatReturnMachine`, but rather `src: someAlreadyInstantiatedMachine` where the Machine is already created as part of boot:

```typescript
// ticketing.ts
const someTicketingMachine = Machine(...);

// userMachineService.ts
const userMachine = Machine(..., {services:{someTicketingMachine}});
```

LMK if this helps

**Reply by davidkpiano** [reply]
There is no difference between an invoke source defined inline and one referenced as a string.

What part are you still confused about? Hopefully I clarified that the thought that "initial events are forwarded" is false.

**Reply by ayashjorden** [reply]
Ok, my observation is that an invoke block with `src: someAsyncFunc` is being invoked with the context and the most recent event.
VS
An invoke block with `src:someSystemMachine` (someSystemMachine is an instance of Machine that was created during boot of my service) receives only the `invoke.start` event and then listens on events sent to it (using `to:invokeBlockId`) and acts on those events.

Considering the following function:
```typescript
async function  fetch(c,e): Promise<User> {
  return await api.getUser(e.userId);
}
```

Is there a difference between the following services? :
```typescript
services: {
 fetch, << plain ref to the function object above
 anotherFetch: (c,e) => fetch(c,e)
}
```

WRT Machines, consider:
```typescript
const notificationsMachine = Machine({...})
```

Should I just use the same const in the `services` object? or `(c,e)=> createNewMachine(c,e)` or anything else?


It might seem like I'm talking about different areas of XState, but I've been playing with it quite a bit and trying to understand when invoke block should use a Machine or a Promise

**Reply by davidkpiano** [reply]
There is no difference to the first two services, and if the machine you're creating doesn't need to care about the `context` or `event`, then yes, you can just pass a machine reference.
